alchemist-engine/src/main/java/it/unibo/alchemist/core/BatchEngine.kt

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright (C) 2010-2023, Danilo Pianini and contributors
+ * listed, for each module, in the respective subproject's build.gradle.kts file.
+ *
+ * This file is part of Alchemist, and is distributed under the terms of the
+ * GNU General Public License, with a linking exception,
+ * as described in the file LICENSE in the Alchemist distribution's top directory.
+ */
+package it.unibo.alchemist.core
+
+import com.google.common.collect.Sets
+import it.unibo.alchemist.boundary.OutputMonitor
+import it.unibo.alchemist.core.model.OutputReplayStrategy
+import it.unibo.alchemist.model.Actionable
+import it.unibo.alchemist.model.Environment
+import it.unibo.alchemist.model.Position
+import it.unibo.alchemist.model.Time
+import kotlinx.coroutines.async
+import kotlinx.coroutines.awaitAll
+import kotlinx.coroutines.runBlocking
+import java.util.function.Consumer
+import java.util.function.Function
+import java.util.stream.Collectors
+
+/**
+ * This class implements a simulation. It offers a wide number of static
+ * factories to ease the creation process.
+ *
+ * @param <T> concentration type
+ * @param <P> [Position] type
+</P></T> */
+class BatchEngine<T, P : Position<out P>> :
```

Comment:
Same as above: can we refactor to favor composition over inheritance?

Reply:
I'm not sure if delegation is applicable in a clean way. The way this class is designed, we override an internal piece of logic doStep().

The only option I see is to extract doStep in a function (Engine<T, P> -> Unit) and bring it into the constructor, however if we use delegation we wouldn't have a way to enforce the implementation of this function, that would leak the logic to the classes constructing an Engine.

Maybe there is some other Kotlin mechanism that would help with this issue I'm missing here?

Reply:
Uhmm... I can see the problem. You would need the delegate to return control, but to do so you need a virtual method, and that method is not declared in the interface.

Which method is being overridden?

Reply:
protected void doStep()

It's been extracted from the main algorithm with the intention to be overriden to customize a single simulation step logic.